Archival

lukasjuhrich opened the pull request:
Prepares #490 
- [x] Discuss whether we want to do this yearly or regularly && adapt code accordingly
- [x] Test archival selection
  - [x] test `select_user_and_last_mem()` select (used as CTE)
  - [x] test cutoff filtering
  - [x] test cutoff calculation from current date
  - [x] test `do-not-archive` filtering
- [ ] ~~Implement `archive_users` (cleanup)~~
- [ ] ~~Add a web endpoint and button archiving all current users~~
- [ ] ~~Add a celerybeat task executing the archival~~

_(update 2026-02-01: in the interest of consolidating changes and moving this forward, we will tackle the follow-up items in separate PRs)_

lukasjuhrich committed:
Add rudimentary archival test

lukasjuhrich committed:
Extract timedelta from archivable membership filter

lukasjuhrich committed:
reintroduce `cast` to silence warning

lukasjuhrich committed:
Add test for longer delta in archivable query

lukasjuhrich committed:
Add `archive_users` stub and test

lukasjuhrich committed:
Extract last membership select from archive query

Introduces `select_users_and_last_mem` returning the `select`

lukasjuhrich committed:
Switch membership selection to `from … select` style

This suits the relational algebra pipeline better, as the entities we
select from actually correspond to a projection done after all the
joins and filtering.

This change does not alter the compilation result of this select statement.

lukasjuhrich committed:
Decouple archivable users select statement from execution

lukasjuhrich committed:
Switch archival query to `from … select` style

lukasjuhrich committed:
Add comment explaining window args

lukasjuhrich committed:
Switch archivable member selection from delta to year-based cutoff

Fixes an oversight in the previous implementation of #470

Also, expand on the test suite substantially while retaining a decent
runtime.

Refs #67

lukasjuhrich committed:
Improve archivability test performance

1. We restrict the ranges to a smaller subset which is still
   meaningful.
2. We use the `select` statement itself rather than the wrapper method
   adding a ton of `joinedload`s.

lukasjuhrich commented:
This PR, as of now, does some base work that can directly be merged into `develop`.
All the rest can be coordinated in one or multiple different PRs (see the extended tasklist of #67).
Waiting for the tests to finish.

georgeto commented on pycroft/lib/user_deletion.py:
```diff
@@ -65,20 +42,59 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
-def get_invalidated_archive_memberships() -> list[Membership]:
-    """Get all memberships in `to_be_archived` of users who have an active `do-not-archive` property.
+    Selected are those users
+    - whose last membership in the member_group ended two weeks in the past,
```
Looks do me like we are now testing for >= 2 years and not weeks.

lukasjuhrich replied:
Fixed in ac859d266f0d

lukasjuhrich committed:
Replace `add_columns` by `with_only_columns` and fix type information

lukasjuhrich committed:
Implement scrubbing routine for `mail`s

This is the first puzzle piece for per-datum scrubbing routines.

lukasjuhrich committed:
Remove `archive_users`

In the new datum-based strategy, this is not the goal anymore.

lukasjuhrich committed:
Implement mail scrubbing in batches of 100

This is a _very_ rudimentary implementation.

lukasjuhrich committed:
Add type hints for `message` in `deferred_gettext`

lukasjuhrich committed:
Add reference to privacy policy to scrubbing functions

lukasjuhrich committed:
Add stubs to other scrubbers with references to privacy policy

FestplattenSchnitzel committed:
Fix typo

FestplattenSchnitzel committed:
Make number of years required for archivability a parameter

lukasjuhrich committed:
Update `psycopg2-binary`, `factory-boy` and add `pytest-freezegun`

lukasjuhrich committed:
model: add `ScrubLog`

lukasjuhrich committed:
include `model.scrubbing` in `model._all`

Otherwise it won't be known to our Metadata

lukasjuhrich committed:
upgrade pytest to v8.4.1

lukasjuhrich committed:
Upgrade uv to v0.7.11

lukasjuhrich committed:
Add a ScrubLogEntry when scrubbing mail address and test scrubbing

lukasjuhrich committed:
Fix tests

lukasjuhrich committed:
Add select statements for other scrubbers

lukasjuhrich committed:
Remove annoying daker args

We want to change the file in local development so we get the fix and
not only the complaint. This is what the default behavior does.

lukasjuhrich committed:
Add stub implementation and test for bulk mail scrubbing

lukasjuhrich committed:
Add bulk mail scrubbing without user log entries

lukasjuhrich committed:
scrublog: Add server_default and fix migration

lukasjuhrich committed:
scrublog: fix indices and nullability in new migration

lukasjuhrich committed:
Fix usage of _WindowArgs type hints

lukasjuhrich committed:
Use `type[]` instead of `typing.Type[]`

lukasjuhrich committed:
scrublog: Use server default for execution time

lukasjuhrich committed:
scrubbing: Allow customization of user filter in bulk mail scrubber

lukasjuhrich committed:
Create common `identity` function in `helpers.functional`

lukasjuhrich committed:
First attempt of a bulk deletion query

FestplattenSchnitzel committed:
Add users with to-be-deleted name, address statements

lukasjuhrich committed:
Add row_number()s to intermediate CTEs so we can properly join

I now officially hate polymorphism in relational databases when based
on joined tables

lukasjuhrich committed:
Replace buggy scrublog migration migrations by proper one

lukasjuhrich committed:
re-lock using `uv` with `--upgrade`

FestplattenSchnitzel committed:
Fix mail deletion query and flash message

lukasjuhrich committed:
Make semgrep happy

lukasjuhrich committed:
Add type annotations to queries

lukasjuhrich committed:
Remove unused import from web.blueprints.user

lukasjuhrich committed:
Use `fetchmany` instead of `partitions` in archivable mails preview

lukasjuhrich committed:
Remove obsolete TODOs

lukasjuhrich committed:
Insert log entries for each removed mail, not just one per scrub

lukasjuhrich committed:
Fix type annotation at membership query

lukasjuhrich committed:
Extend bulk scrubbing test to log entries

lukasjuhrich committed:
Add missing type annotation

lukasjuhrich committed:
Clean up imports

lukasjuhrich committed:
Fix docstring of get_archivable_users

lukasjuhrich commented on pycroft/lib/user_deletion.py:
```diff
@@ -76,20 +67,278 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
+    :param session:
+    :param current_year: dependency injection of the current year.
+    :param years_following_eom: dependency injection of the current year.
+        defaults to the current year.
```
fixed

lukasjuhrich commented on tests/lib/scrubbing/conftest.py:
```diff
@@ -0,0 +1,74 @@
+    Create a user with a membership in a group that has the do-not-archive
+    property.
+    """
+    # TODO wayyy better example set! instead of making all `do-not-archive`,
```
wontfix ;)

lukasjuhrich commented on web/blueprints/user/__init__.py:
```diff
@@ -1492,8 +1500,30 @@ def resend_confirmation_mail() -> ResponseReturnValue:
+
+    sess.commit()
+
+    flash(f"{len(ids.all())} E-Mail-Adressen gelöscht.", "success")
```
I'm confused by this: since when are `ScalarResult`s materialized? they still have `fetchmany`, `all` as methods, and  for good reason. The latter calls `ResultInternal._allrows()`.

lukasjuhrich replied:
wontfix.

lukasjuhrich committed:
Merge develop and archival

lukasjuhrich commented on web/blueprints/user/__init__.py:
```diff
@@ -1492,8 +1500,30 @@ def resend_confirmation_mail() -> ResponseReturnValue:
+        "user/archivable_users.html",
+        table=table,
+        archivable_mail_count=scrubbable_mails_count(sess, datetime.now().year),
+        archivable_mail_users=next(scrubbable_mails(sess).partitions(10), []),
```
fixd

lukasjuhrich commented on tests/lib/scrubbing/conftest.py:
```diff
@@ -0,0 +1,74 @@
+
+@pytest.fixture(scope="module")
+def user_archivable(module_session: Session, config: Config) -> User:
+    # TODO: user wtih membership
```
removed

lukasjuhrich commented on tests/lib/scrubbing/conftest.py:
```diff
@@ -0,0 +1,74 @@
+
+@pytest.fixture(scope="module")
+def users_archivable(module_session: Session, config: Config) -> list[User]:
+    # TODO: user wtih membership
```
removed

lukasjuhrich commented on pycroft/lib/membership.py:
```diff
@@ -239,3 +243,46 @@ def change_membership_active_during(
     log_user_event(message, processor, membership.user)
+
+
+def select_user_and_last_mem() -> Select[tuple[int, int, str]]:
```
fixed

lukasjuhrich committed:
Re-Lock

lukasjuhrich committed:
Update uv to 0.9.28

lukasjuhrich commented on pycroft/lib/user_deletion.py:
```diff
@@ -76,20 +67,274 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
+    author: User,
+    user_filter: t.Callable[[Select[tuple[User]]], Select[tuple[User]]] | None = None,
+) -> ScalarResult[int]:
+    return session.scalars(scrub_all_mails_stmt(utcnow().year, author.id, user_filter))
```
wrong, this is our `utcnow()` helper.

lukasjuhrich commented on pycroft/lib/user_deletion.py:
```diff
@@ -76,20 +67,278 @@ def get_archivable_members(session: Session) -> Sequence[ArchivableMemberInfo]:
+        )
+        .cte("user_log_entries")
+    )
+    return select(removed_mail_user_ids.c.id).add_cte(insert_scrub_log, insert_user_log)
```
AFAIK this is not true in postgres, and tests also indicate that they are executed properly.

lukasjuhrich commented on tests/lib/scrubbing/conftest.py:
```diff
@@ -0,0 +1,72 @@
+    Create a user with a membership in a group that has the do-not-archive
+    property.
+    """
+    # TODO wayyy better example set! instead of making all `do-not-archive`,
+    # we would want them to have different membership ranges, also additional memberships;
+    # also edge-cases like upcoming memberships or on-new-years membership terminations
```
Sure, but that is not a blocker for me.

lukasjuhrich commented on web/blueprints/user/__init__.py:
```diff
@@ -9,9 +9,10 @@
 
     :copyright: (c) 2012 by AG DSN.
 """
+from dataclasses import dataclass
```
removed.
